cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc

```diff
@@ -273,12 +273,27 @@ void OptimizationProblem2D::Solve(
         continue;
       }
 
-      const transform::Rigid3d relative_pose =
-          ComputeRelativePose(trajectory_id, first_node_data, second_node_data);
+      // Add a relative pose constraint based on the odometry (if available).
+      std::unique_ptr<transform::Rigid3d> relative_odometry =
+          OdometryBetween(trajectory_id, first_node_data, second_node_data);
+      if (relative_odometry != nullptr) {
+        problem.AddResidualBlock(
+            SpaCostFunction2D::CreateAutoDiffCostFunction(Constraint::Pose{
+                (*relative_odometry), options_.odometry_translation_weight(),
+                options_.odometry_rotation_weight()}),
+            nullptr /* loss function */, C_nodes.at(first_node_id).data(),
+            C_nodes.at(second_node_id).data());
+      }
+
+      // Add a relative pose constraint based on consecutive local SLAM poses.
+      const transform::Rigid3d relative_local_slam_pose =
+          transform::Embed3D(first_node_data.initial_pose.inverse() *
+                             second_node_data.initial_pose);
       problem.AddResidualBlock(
-          SpaCostFunction2D::CreateAutoDiffCostFunction(Constraint::Pose{
-              relative_pose, options_.consecutive_node_translation_weight(),
-              options_.consecutive_node_rotation_weight()}),
+          SpaCostFunction2D::CreateAutoDiffCostFunction(
+              Constraint::Pose{relative_local_slam_pose,
+                               options_.local_slam_pose_translation_weight(),
+                               options_.local_slam_pose_rotation_weight()}),
```

Comment:
Is Ceres smart enough not to include constraints in the optimization problem that have a weight of 0? So I assume we don't need to special case this.

Reply:
Hmm... you mean if it still works if someone sets the weights to zero? Not sure.

Reply:
As far as I can tell, the weights are not known to Ceres, they are applied in our custom cost functor [here](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping/internal/2d/pose_graph/spa_cost_function_2d.h#L51), specifically [here](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping/internal/pose_graph/cost_helpers_impl.h#L42).

Reply:
Yes, that is what I was wondering about.